# Pandas stands for Python Data Analysis Library

# What is Library ? 

# Library is actually a set of predefined functions developed to achieve a specific class of operations

# Now, using Pandas Library, you actually work on data and get valuable descriptive as well inferential statistics from your data

# The types of data which we will be dealing with in PGDIDS&ML is going to be four types:

# Tabular Data (In the form of tables, excel files, csv file,.......etc)

# Textual Data (In the form of documents, word files, pdfs, notepad files...etc)

# Images Data (In the form of jpeg, png files)

# Videos (In the form of mp4 or whatever)

# Pandas library is primarily used to work on tabular data

# CSV stands for Comma Seperated Value Format. It is just like your excel format. 

# If we want to do any kind of analysis on csv data then we need to first of all bring that data into python environment and for that we have to read the data, that means we have to first fetch everything in tabular data format file. Now, let's see how can we do that: 

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("mushrooms.csv")

In [3]:
df

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
5        e         x           y         y       t    a               f   
6        e         b           s         w       t    a               f   
7        e         b           y         w       t    l               f   
8        p         x           y         w       t    p               f   
9        e         b           s         y       t    a               f   
10       e         x           y         y       t    l               f   
11       e         x           y         y       t    a               f   
12       e         b           s         y       t    a               f   
13       p         x           y         w       t    p               f   
14       e         x           f         n       f    n               f   
15       e         s           f         g       f    n               f   
16       e         f           f         w       f    n               f   
17       p         x           s         n       t    p               f   
18       p         x           y         w       t    p               f   
19       p         x           s         n       t    p               f   
20       e         b           s         y       t    a               f   
21       p         x           y         n       t    p               f   
22       e         b           y         y       t    l               f   
23       e         b           y         w       t    a               f   
24       e         b           s         w       t    l               f   
25       p         f           s         w       t    p               f   
26       e         x           y         y       t    a               f   
27       e         x           y         w       t    l               f   
28       e         f           f         n       f    n               f   
29       e         x           s         y       t    a               f   
...    ...       ...         ...       ...     ...  ...             ...   
8094     e         b           s         g       f    n               f   
8095     p         x           y         c       f    m               f   
8096     e         k           f         w       f    n               f   
8097     p         k           y         n       f    s               f   
8098     p         k           s         e       f    y               f   
8099     e         k           f         w       f    n               f   
8100     e         f           s         n       f    n               a   
8101     p         k           s         e       f    s               f   
8102     e         x           s         n       f    n               a   
8103     e         k           s         n       f    n               a   
8104     e         k           s         n       f    n               a   
8105     e         k           s         n       f    n               a   
8106     e         k           s         n       f    n               a   
8107     e         x           s         n       f    n               a   
8108     p         k           y         e       f    y               f   
8109     e         b           s         w       f    n               f   
8110     e         x           s         n       f    n               a   
8111     e         k           s         w       f    n               f   
8112     e         k           s         n       f    n               a   
8113     p         k           y         e       f    y               f   
8114     p         f           y         c       f    m               a   
8115     e         x          

# The way your data is getting displayed above is a different way of storing the data apart from lists, dictionaries, tuples and sets. This way of storing the data is Pandas DataFrame. A Pandas DataFrame is another kind of Data Structure. 

In [6]:
df.iloc[7:13]

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
7      e         b           y         w       t    l               f   
8      p         x           y         w       t    p               f   
9      e         b           s         y       t    a               f   
10     e         x           y         y       t    l               f   
11     e         x           y         y       t    a               f   
12     e         b           s         y       t    a               f   

   gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
7             c         b          n  ...                        s   
8             c         n          p  ...                        s   
9             c         b          g  ...                        s   
10            c         b          g  ...                        s   
11            c         b          n  ...                        s   
12            c         b          w  ...                        s   

   stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
7                       w                      w         p          w   
8                       w                      w         p          w   
9                       w                      w         p          w   
10                      w                      w         p          w   
11                      w                      w         p          w   
12                      w                      w         p          w   

   ring-number ring-type spore-print-color population habitat  
7            o         p                 n          s       m  
8            o         p                 k          v       g  
9            o         p                 k          s       m  
10           o         p                 n          n       g  
11           o         p                 k          s       m  
12           o         p                 n          s       g  

[6 rows x 23 columns]

# In the above manner, you can access rows of the Pandas DataFrame using iloc function

# Now let's see that how can we access columns of a Pandas DataFrame

In [7]:
df

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
5        e         x           y         y       t    a               f   
6        e         b           s         w       t    a               f   
7        e         b           y         w       t    l               f   
8        p         x           y         w       t    p               f   
9        e         b           s         y       t    a               f   
10       e         x           y         y       t    l               f   
11       e         x           y         y       t    a               f   
12       e         b           s         y       t    a               f   
13       p         x           y         w       t    p               f   
14       e         x           f         n       f    n               f   
15       e         s           f         g       f    n               f   
16       e         f           f         w       f    n               f   
17       p         x           s         n       t    p               f   
18       p         x           y         w       t    p               f   
19       p         x           s         n       t    p               f   
20       e         b           s         y       t    a               f   
21       p         x           y         n       t    p               f   
22       e         b           y         y       t    l               f   
23       e         b           y         w       t    a               f   
24       e         b           s         w       t    l               f   
25       p         f           s         w       t    p               f   
26       e         x           y         y       t    a               f   
27       e         x           y         w       t    l               f   
28       e         f           f         n       f    n               f   
29       e         x           s         y       t    a               f   
...    ...       ...         ...       ...     ...  ...             ...   
8094     e         b           s         g       f    n               f   
8095     p         x           y         c       f    m               f   
8096     e         k           f         w       f    n               f   
8097     p         k           y         n       f    s               f   
8098     p         k           s         e       f    y               f   
8099     e         k           f         w       f    n               f   
8100     e         f           s         n       f    n               a   
8101     p         k           s         e       f    s               f   
8102     e         x           s         n       f    n               a   
8103     e         k           s         n       f    n               a   
8104     e         k           s         n       f    n               a   
8105     e         k           s         n       f    n               a   
8106     e         k           s         n       f    n               a   
8107     e         x           s         n       f    n               a   
8108     p         k           y         e       f    y               f   
8109     e         b           s         w       f    n               f   
8110     e         x           s         n       f    n               a   
8111     e         k           s         w       f    n               f   
8112     e         k           s         n       f    n               a   
8113     p         k           y         e       f    y               f   
8114     p         f           y         c       f    m               a   
8115     e         x          

In [12]:
df_columns = list(df.columns)

In [27]:
df_columns

['class',
 'cap-shape',
 'cap-surface',
 'cap-color',
 'bruises',
 'odor',
 'gill-attachment',
 'gill-spacing',
 'gill-size',
 'gill-color',
 'stalk-shape',
 'stalk-root',
 'stalk-surface-above-ring',
 'stalk-surface-below-ring',
 'stalk-color-above-ring',
 'stalk-color-below-ring',
 'veil-type',
 'veil-color',
 'ring-number',
 'ring-type',
 'spore-print-color',
 'population',
 'habitat']

In [20]:
df['cap-surface']

0       s
1       s
2       s
3       y
4       s
5       y
6       s
7       y
8       y
9       s
10      y
11      y
12      s
13      y
14      f
15      f
16      f
17      s
18      y
19      s
20      s
21      y
22      y
23      y
24      s
25      s
26      y
27      y
28      f
29      s
       ..
8094    s
8095    y
8096    f
8097    y
8098    s
8099    f
8100    s
8101    s
8102    s
8103    s
8104    s
8105    s
8106    s
8107    s
8108    y
8109    s
8110    s
8111    s
8112    s
8113    y
8114    y
8115    s
8116    y
8117    s
8118    y
8119    s
8120    s
8121    s
8122    y
8123    s
Name: cap-surface, Length: 8124, dtype: object

In [26]:
df.loc[7:12,'cap-surface':'gill-color']

cap-surface cap-color bruises odor gill-attachment gill-spacing gill-size  \
7            y         w       t    l               f            c         b   
8            y         w       t    p               f            c         n   
9            s         y       t    a               f            c         b   
10           y         y       t    l               f            c         b   
11           y         y       t    a               f            c         b   
12           s         y       t    a               f            c         b   

   gill-color  
7           n  
8           p  
9           g  
10          g  
11          n  
12          w

In [28]:
df.iloc[7:13,2:10]

cap-surface cap-color bruises odor gill-attachment gill-spacing gill-size  \
7            y         w       t    l               f            c         b   
8            y         w       t    p               f            c         n   
9            s         y       t    a               f            c         b   
10           y         y       t    l               f            c         b   
11           y         y       t    a               f            c         b   
12           s         y       t    a               f            c         b   

   gill-color  
7           n  
8           p  
9           g  
10          g  
11          n  
12          w

# Suppose, I want to compute the following probability:

\begin{equation}
P(class = 'p')
\end{equation}

# Well, I have a sample of 8124 Mushrooms, but I don't have the population. Therefore, I can calculate the estimate of this probability using the sample. And the estimate is given by the relative frequency. The relative frequency of Poissonous mushrooms is given by: 

\begin{equation}
\hat{P}(class = poissonous) = \frac{Frequency(Mushroom = Poissonous)}{Total Number of Mushrooms (Sample Size)}
\end{equation}

\begin{equation}
\hat{P}(class = poissonous) = \frac{Number of rows corresponsing to poissonous mushrooms in the Pandas DataFrame}{Sample Size (Total Number of rows in Pandas DataFrame)}
\end{equation}

# So, how we are going to compute the Frequency of poissonous mushrooms in our data. For that, we have to seperate the data of all the poissonous mushrooms from our Pandas DataFrame. This seperated data will be in the form of another smaller DataFrame. Then we will be computing the total number of rows in the smaller DataFrame. 

# Now, how we are going to do that ? 

# In order to understand this, we first have to understand the concept of Boolean masks. 

# So, what are Booleans Masks ? 

# Boolean Masks are actually the DataFrame of Boolean values (True and False) where True corresponds to some condition being True and False corresponds to the same condition being False. 

# Now, the next concept after slicing of DataFrame from iloc as well as loc, is going to be of Boolean Masks. 

In [29]:
df_columns[0]

'class'

In [30]:
df['class']

0       p
1       e
2       e
3       p
4       e
5       e
6       e
7       e
8       p
9       e
10      e
11      e
12      e
13      p
14      e
15      e
16      e
17      p
18      p
19      p
20      e
21      p
22      e
23      e
24      e
25      p
26      e
27      e
28      e
29      e
       ..
8094    e
8095    p
8096    e
8097    p
8098    p
8099    e
8100    e
8101    p
8102    e
8103    e
8104    e
8105    e
8106    e
8107    e
8108    p
8109    e
8110    e
8111    e
8112    e
8113    p
8114    p
8115    e
8116    p
8117    p
8118    p
8119    e
8120    e
8121    e
8122    p
8123    e
Name: class, Length: 8124, dtype: object

In [32]:
boolean_mask = df['class'] == 'p'

# Now, as you can that the above DataFrame is a DataFrame of Boolean Values and hence it is called boolean mask. 

# Now, if we pass this boolean mask inside the square brackets of our pandas dataframe:

# df[boolean_mask]

# then, let's see what happens:

In [34]:
poissonous_mushrooms = df[boolean_mask]

# we have to find now out of this data that, how many mushrooms are there which have poissonous category so for that, we have a function associated with the DataFrame and that is called shape. 

In [40]:
poissonous_mushrooms.shape[0]

3916

# So, finally we were able to find the Frequency(Mushroom = Poissonous) = 3916

# So, we have to compute now the total number of mushrooms (sample size) in our data

In [41]:
sample_size = df.shape[0]

In [42]:
sample_size

8124

In [43]:
p_class_equals_poissonous = poissonous_mushrooms.shape[0]/df.shape[0]

In [44]:
p_class_equals_poissonous

0.48202855736090594

# Therefore, 

\begin{equation}
\hat{P}(Class = Poissonous) = 0.48202855736090594
\end{equation}